File: src/test/java/kitchenpos/product/infra/persistence/FakeProductRepository.java
package kitchenpos.product.infra.persistence;

import kitchenpos.product.domain.model.Product;
import kitchenpos.product.domain.repository.ProductRepository;

import java.util.*;

public class FakeProductRepository implements ProductRepository {

    private final Map<UUID, Product> storage;

    public FakeProductRepository(Map<UUID, Product> storage) {
        this.storage = storage;
    }

    @Override
    public List<Product> findAllByIdIn(List<UUID> ids) {
        return new ArrayList<>(storage.values().stream()
                .filter(p -> ids.contains(p.getId()))
                .toList());
    }

    @Override
    public Optional<Product> findById(UUID id) {
        return Optional.ofNullable(storage.get(id));
    }

    @Override
    public Product save(Product product) {
        UUID id = UUID.randomUUID();
        product.setId(id);
        storage.put(id, product);
        return product;
    }

    @Override
    public List<Product> findAll() {
        return new ArrayList<>(storage.values());
    }
}

File: src/test/kotlin/kitchenpos/product/tobe/infra/FakeProductRepository.kt
package kitchenpos.product.tobe.infra

import kitchenpos.product.tobe.domain.Product
import kitchenpos.product.tobe.domain.ProductRepository
import java.util.*

class FakeProductRepository(
    private val products: MutableMap<UUID, Product> = mutableMapOf()

) : ProductRepository {

    override fun save(product: Product): Product {
        if (product.id == null) {
            product.id = UUID.randomUUID()
        }
        products[product.id!!] = product
        return product
    }

    override fun findById(id: UUID): Optional<Product> {
        return Optional.ofNullable(products[id])
    }

    override fun findAll(): List<Product> {
        return products.values.toList()
    }

    override fun findAllByIdIn(ids: List<UUID>): List<Product> {
        return ids.map { products[it]!! }
    }
}
